src/recommendation/src/main/scala/RankingEvaluator.scala

```diff
@@ -0,0 +1,162 @@
+        reciprocalRank
+      } else {
+        0.0
+      }
+    }.mean()
+  }
+  lazy val fractionConcordantPairs: Double = {
+    predictionAndLabels.map { case (pred, lab) =>
+      var nc = 0.0
+      var nd = 0.0
+      pred.zipWithIndex.foreach(a => {
+        if (lab.length > a._2) {
+          if (a._1 == lab(a._2)) nc += 1
+          else nd += 1
+        }
+      })
+      nc / (nc + nd)
+    }.mean()
+  }
+
+  def matchMetric(metricName: String): Double = metricName match {
+    case "map"          => map
+    case "ndcgAt"       => ndcg
+    case "precisionAtk" => precisionAtk
+    case "recallAtK"    => recallAtK
+    case "diversityAtK" => diversityAtK
+    case "maxDiversity" => maxDiversity
+    case "mrr"          => meanReciprocalRank
+    case "fcp"          => fractionConcordantPairs
+  }
+
+  def getAllMetrics: Map[String, Double] = {
+    Map("map" -> map,
+      "ndcgAt" -> ndcg,
+      "precisionAtk" -> precisionAtk,
+      "recallAtK" -> recallAtK,
+      "diversityAtK" -> diversityAtK,
+      "maxDiversity" -> maxDiversity,
+      "mrr" -> meanReciprocalRank,
+      "fcp" -> fractionConcordantPairs)
+  }
+}
+
+class RankingEvaluator(override val uid: String)
+  extends Evaluator with RecEvaluatorParams with HasRecommenderCols with ComplexParamsWritable {
+
+  def this() = this(Identifiable.randomUID("recEval"))
+
+  val nItems: LongParam = new LongParam(this, "nItems", "number of items")
+
+  def setNItems(value: Long): this.type = set(nItems, value)
+
+  def getNItems: Long = $(nItems)
+
+  val k: IntParam = new IntParam(this, "k",
+    "number of items", ParamValidators.inRange(1, Integer.MAX_VALUE))
+  setDefault(k -> 1)
+
+  /** @group getParam */
+  def getK: Int = $(k)
+
+  /** @group setParam */
+  def setK(value: Int): this.type = set(k, value)
+
+  val metricName: Param[String] = {
+    val allowedParams = ParamValidators.inArray(Array("ndcgAt", "map", "mapk", "recallAtK", "diversityAtK",
+      "maxDiversity", "mrr", "fcp"))
+    new Param(this, "metricName", "metric name in evaluation " +
+      "(ndcgAt|map|precisionAtk|recallAtK|diversityAtK|maxDiversity|mrr|fcp)",
+      allowedParams)
+  }
+  setDefault(metricName -> "ndcgAt", nItems -> -1)
+
+  /** @group getParam */
+  def getMetricName: String = $(metricName)
+
+  /** @group setParam */
+  def setMetricName(value: String): this.type = set(metricName, value)
+
+  /** @group setParam */
+  def setLabelCol(value: String): this.type = set(labelCol, value)
+
+  /** @group setParam */
+  def setPredictionCol(value: String): this.type = set(predictionCol, value)
+
+  def getMetrics(dataset: Dataset[_]): AdvancedRankingMetrics[Any] = {
+    val predictionAndLabels = dataset
+      .select(getPredictionCol, getLabelCol)
+      .rdd.map { case Row(prediction: Seq[Any], label: Seq[Any]) => (prediction.toArray, label.toArray) }
+      .cache()
+    new AdvancedRankingMetrics[Any](predictionAndLabels, getK, getNItems)
+    //TODO come back to this Any type
```

Comment:
why is any type needed here?

Reply:
When I try to switch everything to Ints (which is what they are instead of any) I get this...
![image](https://user-images.githubusercontent.com/9027725/47731633-e3088380-dc3a-11e8-99f2-60ed386ecd59.png)